### ISSUE TITLE
failed to parse function return function pointers

### ISSUE THREAD

**Issue by minux** [rui314/8cc#64]
```
void (*f(int *p))(void) {
	return (void *)0;
}
```

```
$ 8cc -c t2.c 
[ERROR] 8cc.h:426: t.c:1:25: ';' or ',' are expected, but got {
```

**Comment by andrewchambers** [comment]
Looks like is_funcdef() in parse.c makes some incorrect assumptions so it tries to parse this as a decl.

**Comment by rui314** [comment]
That's right. is_funcdef returns a wrong result for the input. Maybe we should stop making a guess before calling the parser but instead call the parser directly, because making a good guess is as hard as parsing input.

**Comment by andrewchambers** [comment]
The look ahead is good for readability. If we don't have the ability to save and reset the parser at choice points on an error,  decl and function parsing may need to be combined into an ugly mess.

**Comment by andrewchambers** [comment]
It actually might not be too bad, I *think* it comes down to a check in read_decl like:
```
decl = read_decl()
if ( peek == '{' && ! first decl in list  ) 
  error("function cannot appear in decl list");
if (peek == '{') { 
  if (sclass == typdef)
    error()
  parse_body() 
} 
```
I think the type struct already parses the vector of param names.
